dracut: relabel the hostname file

yuqi-zhang opened the pull request:
Mark the file as needing relabelling, otherwise the hostname
file written to by the afterburn-hostname.service will get
SELinux denials.

This is somewhat of a hack. Eventually we want to relabel in the initrd. See: https://github.com/coreos/coreos-assembler/pull/551#issuecomment-500852247

cc @jlebon

jlebon commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -16,4 +16,6 @@ OnFailureJobMode=isolate
 [Service]
 ExecStart=/usr/bin/afterburn --cmdline --hostname=/sysroot/etc/hostname
+ExecStart=/bin/sh -c 'mkdir -p /sysroot/run/tmpfiles.d'
+ExecStart=/bin/sh -c 'echo "Z /etc/hostname - - -" > /sysroot/run/tmpfiles.d/hostname-relabel.conf'
```
Minor: just needs `z` there. `Z` is for recursive labeling.

jlebon commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -16,4 +16,6 @@ OnFailureJobMode=isolate
 
 [Service]
 ExecStart=/usr/bin/afterburn --cmdline --hostname=/sysroot/etc/hostname
+ExecStart=/bin/sh -c 'mkdir -p /sysroot/run/tmpfiles.d'
```
This should be just `/run/tmpfiles.d`. The `/run` tmpfs is carried over from the initrd at switchroot.

jlebon commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -16,4 +16,6 @@ OnFailureJobMode=isolate
 
 [Service]
 ExecStart=/usr/bin/afterburn --cmdline --hostname=/sysroot/etc/hostname
+ExecStart=/bin/sh -c 'mkdir -p /sysroot/run/tmpfiles.d'
```
Also, maybe throw a comment to e.g. https://github.com/coreos/ignition/issues/635? (Or a more targeted issue if there is one).

yuqi-zhang replied:
Will do

jlebon replied:
Heh, I meant a comment in this file to that GitHub issue, but consider it optional.

jlebon commented on dracut/30afterburn/afterburn-hostname.service:
```diff
@@ -16,4 +16,6 @@ OnFailureJobMode=isolate
 [Service]
 ExecStart=/usr/bin/afterburn --cmdline --hostname=/sysroot/etc/hostname
+ExecStart=/bin/sh -c 'mkdir -p /sysroot/run/tmpfiles.d'
+ExecStart=/bin/sh -c 'echo "Z /etc/hostname - - -" > /sysroot/run/tmpfiles.d/hostname-relabel.conf'
```
You'll likely also need to order whatever problematic services to *after* `systemd-tmpfiles-setup.service`. See e.g. https://github.com/coreos/fedora-coreos-config/blob/1842e9e309c2dfc89863705917c19f8eb85be245/overlay/usr/lib/dracut/modules.d/40coreos-var/coreos-populate-var.sh#L39. Depending on the service, that might be tricky to do.

yuqi-zhang replied:
Shouldn't be an issue since AFAIK only NM should be using this

jlebon replied:
Hmm, does NM have an explicit ordering to `systemd-tmpfiles-setup.service`?

jlebon replied:
Well, none of the NM units use `DefaultDependencies=no`, so it should already be ordered after `sysinit.target` (which pulls in `systemd-tmpfiles-setup.service`), so yeah, I think we're fine.

jlebon reviewed (commented):
At least for Fedora, you could use systemd's `/run/systemd/relabel-extra.d/`: https://github.com/systemd/systemd/pull/11131. Though there's value in keeping the same workaround for FCOS and RHCOS too.

jlebon reviewed (approved):
LGTM. Optional suggestion to add a comment, but good as is too.

yuqi-zhang committed:
dracut: relabel the hostname file

Mark the file as needing relabelling, otherwise the hostname
file written to by the afterburn-hostname.service will get
SELinux denials.

Signed-off-by: Yu Qi Zhang <jerzhang@redhat.com>
